src/wrappers/fastify.js

```diff
@@ -0,0 +1,148 @@
+    tracer.sendTrace(() => {}, tracerObj).then(resolve).then(() => {
+        utils.debugLog('[fastify] - trace sent + request resolved');
+    });
+}
+
+/**
+ * Fastify requests middleware that runs in context
+ * @param {Request} request The Fastify's request data
+ * @param {Response} reply The Fastify's response data
+ */
+function fastifyMiddleware(request, reply) {
+    // Check if endpoint is ignored
+    traceContext.setMainReference();
+    utils.debugLog('[fastify] - starting middleware');
+    const tracerObj = tracer.getTrace();
+    if (!tracerObj) {
+        utils.debugLog('[fastify] - no tracer found on init');
+    }
+    if (shouldIgnore(request.url, request.headers)) {
+        utils.debugLog(`Ignoring request: ${request.url}`);
+        return;
+    }
+    const chunks = [];
+    request.raw.on('data', chunk => httpHelpers.addChunk(chunk, chunks));
+
+    tracer.restart();
+    let fastifyEvent;
+    const startTime = Date.now();
+    try {
+        fastifyEvent = fastifyRunner.createRunner(request, startTime);
+        utils.debugLog('[fastify] - created runner');
+        // Handle response
+        const requestPromise = new Promise((resolve) => {
+            traceContext.setAsyncReference(tracerObj);
+            utils.debugLog('[fastify] - creating response promise');
+            request.raw.once('end', () => handleResponse(
+                tracerObj,
+                fastifyEvent,
+                reply,
+                request,
+                startTime,
+                resolve,
+                Buffer.concat(chunks).toString()
+            ));
+        });
+
+        request.context._EPSAGON_EVENT = fastifyEvent;
+        if (!request.context._originalErrorHandler) {
+            request.context._originalErrorHandler = request.context.errorHandler;
+            request.context.errorHandler = (err, req, rep) => {
+                eventInterface.setException(req.context._EPSAGON_EVENT, err);
+                request.context._originalErrorHandler(err, req, rep);
+            };
+        }
+
+        tracer.addRunner(fastifyEvent, requestPromise);
+        utils.debugLog('[fastify] - added runner');
+
+        // Inject trace functions
+        const { label, setError, getTraceUrl } = tracer;
+        request.epsagon = {
+            label,
+            setError,
+            getTraceUrl,
+        };
+        traceContext.setMainReference(false);
+    } catch (err) {
+        utils.debugLog('[fastify] - general catch');
+        utils.debugLog(err);
+    } finally {
+        utils.debugLog('[fastify] - general finally');
+    }
+}
+
+/**
+ * Wraps the Fastify module request function with tracing
+ * @param {Function} wrappedFunction Fastify init function
+ * @return {Function} updated wrapped init
+ */
+function fastifyWrapper(wrappedFunction) {
+    utils.debugLog('[fastify] - wrapping');
+    return function internalFastifyWrapper(functions, runner, request, reply, cb) {
```

Comment:
Add failsafe behaviour in case the internal code fails. e.g (fastifyMiddleware throws an exception, we'd want to run the original code)

Reply:
done 👍🏻